shared-module/uvc/__init__.c

```diff
@@ -0,0 +1,198 @@
+#include "py/runtime.h"
+#include <stdint.h>
+#include "class/video/video_device.h"
+#include "shared-bindings/displayio/Bitmap.h"
+#include "shared-bindings/uvc/__init__.h"
+#include "shared-module/bitmapfilter/macros.h"
+#include "shared-module/uvc/__init__.h"
+#include "shared-module/uvc/uvc_usb_descriptors.h"
+#include "supervisor/background_callback.h"
+#include "supervisor/shared/tick.h"
+#include "device/usbd.h"
+
+static bool do_convert = true;
+static unsigned frame_num = 0;
+static unsigned tx_busy = 0;
+static unsigned interval_ms = 1000 / DEFAULT_FRAME_RATE;
+
+// TODO must dynamically allocate this, otherwise everyone pays for it
+static uint8_t *frame_buffer_yuyv;
+uint16_t *uvc_framebuffer_rgb565;
+
+static bool uvc_is_enabled = false;
+uint16_t uvc_frame_width, uvc_frame_height;
+
+bool shared_module_uvc_enable(mp_int_t frame_width, mp_int_t frame_height) {
+    if (tud_connected()) {
+        return false;
+    }
+
+    // this will free any previously allocated framebuffer as a side-effect
+    shared_module_uvc_disable();
+
+    if (frame_width & 1) {
+        // frame_width must be even, round it up
+        frame_width++;
+    }
+
+    uvc_frame_width = frame_width;
+    uvc_frame_height = frame_height;
+
+    size_t framebuffer_size = uvc_frame_width * uvc_frame_height * 2;
+    frame_buffer_yuyv = port_malloc(framebuffer_size, false);
```

Comment:
Does ESP work if you set the dma bool to true?

Reply:
No, I had the same thought and tried changing that flag to true.